planner/core/find_best_task.go

```diff
@@ -279,13 +279,48 @@ func (p *baseLogicalPlan) enumeratePhysicalPlans4Task(physicalPlans []PhysicalPl
 		}
 		opt.appendCandidate(p, curTask.plan(), prop)
 		// Get the most efficient one.
-		if curTask.cost() < bestTask.cost() || (bestTask.invalid() && !curTask.invalid()) {
+		if curIsBetter, err := compareTask(p.ctx, curTask, bestTask); err != nil {
+			return nil, 0, err
+		} else if curIsBetter {
 			bestTask = curTask
 		}
 	}
 	return bestTask, cntPlan, nil
 }
 
+func compareTask(ctx sessionctx.Context, curTask, bestTask task) (curIsBetter bool, err error) {
+	if ctx.GetSessionVars().EnableNewCostInterface { // use the new cost interface
+		curCost, err := getTaskPlanCost(curTask)
+		if err != nil {
+			return false, err
+		}
+		bestCost, err := getTaskPlanCost(bestTask)
+		if err != nil {
+			return false, err
+		}
+		return curCost < bestCost || (bestTask.invalid() && !curTask.invalid()), nil
+	}
+	return curTask.cost() < bestTask.cost() || (bestTask.invalid() && !curTask.invalid()), nil
+}
+
+func getTaskPlanCost(t task) (float64, error) {
+	if t.invalid() {
+		return math.MaxFloat64, nil
+	}
+	var taskType property.TaskType
+	switch t.(type) {
+	case *rootTask:
+		taskType = property.RootTaskType
+	case *copTask:
+		taskType = property.CopSingleReadTaskType
+	case *mppTask:
+		taskType = property.MppTaskType
+	default:
+		return 0, errors.New("unknown task type")
```

Comment:
Where is `CopDoubleReadTaskType`?

Reply:
`CopSingleReadTaskType` has the same meaning as `CopSingleReadTaskType` here since when calculating cost we don't need to know whether the task is `single-read` or `double-read`. I'll add some comment here.

Reply:
Some comments were added, PTAL.